### PR TITLE
Prepare AllocateLoadBalancerNodePorts for GA

### PR DESCRIPTION
**Pull request by uablrek**
#### What type of PR is this?

Prepares the `AllocateLoadBalancerNodePorts` feature for GA.  Feature-gate: `ServiceLBNodePortControl`.

/kind cleanup

#### What this PR does / why we need it:

To allow AllocateLoadBalancerNodePorts to graduate.

#### Which issue(s) this PR fixes:

N/A

#### Does this PR introduce a user-facing change?

```release-note
The `ServiceLBNodePortControl` feature graduates to GA. The feature gate will be removed in 1.26.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1864-disable-lb-node-ports)

### CODE REVIEW

**Comment by uablrek** [comment]
There are `ServiceLBNodePortControl` still comments in 2 files that seem to be generated;
* staging/src/k8s.io/cli-runtime/artifacts/openapi/swagger.json
* staging/src/k8s.io/kubectl/testdata/openapi/swagger.json

I have executed `hack/update-generated-swagger-docs.sh` but these are not updated.

**Review by andrewsykim — COMMENTED** [review]
Overall LGTM, thanks for taking this to GA @uablrek!

File: pkg/registry/core/service/strategy.go [hunk 1]
```diff
@@ -162,12 +162,6 @@ func (svcStrategy) AllowUnconditionalUpdate() bool {
 //         newSvc.Spec.MyFeature = nil
 //     }
 func dropServiceDisabledFields(newSvc *api.Service, oldSvc *api.Service) {
-	// Clear AllocateLoadBalancerNodePorts if ServiceLBNodePortControl is not enabled
-	if !utilfeature.DefaultFeatureGate.Enabled(features.ServiceLBNodePortControl) {
-		if !allocateLoadBalancerNodePortsInUse(oldSvc) {
```
Comment (aojea) [inline pkg/registry/core/service/strategy.go:167]: this function is dead code after removing this
```
// returns true if svc.Spec.AllocateLoadBalancerNodePorts field is in use
func allocateLoadBalancerNodePortsInUse(svc *api.Service) bool {
	if svc == nil {
		return false
	}
	return svc.Spec.AllocateLoadBalancerNodePorts != nil
}
```
Line 193

File: pkg/registry/core/service/storage/alloc.go [hunk 2]
```diff
@@ -940,10 +938,7 @@ func shouldAllocateNodePorts(service *api.Service) bool {
 		return true
 	}
 	if service.Spec.Type == api.ServiceTypeLoadBalancer {
-		if utilfeature.DefaultFeatureGate.Enabled(features.ServiceLBNodePortControl) {
-			return *service.Spec.AllocateLoadBalancerNodePorts
-		}
-		return true
+		return *service.Spec.AllocateLoadBalancerNodePorts
```
Comment (aojea) [inline pkg/registry/core/service/storage/alloc.go:941]: @thockin my memory is blur here, it is working but I'd  like to fully understand it.

shouldAllocateNodePorts() is called from updateNodePorts() or initNodePorts().

beginCreate() -> allocateCreate() -> txnAllocNodePorts() -> initNodePorts()  
beginUpdate() allocateUpdate() -> txnUpdateNodePorts() -> updateNodePorts()

Those hooks run at the beginning, is the defaulting happening before?

https://github.com/kubernetes/kubernetes/blob/fbdd0d7b4165bc5a677d45e4dc693e3260297bfa/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L366-L386

File: pkg/registry/core/service/strategy.go (hunk 1 above)
Comment (uablrek) [reply pkg/registry/core/service/strategy.go:167]: Yeah, `golangci-lint` saw that too :smiley:

**Comment by uablrek** [comment]
/remove-sig apps
(again? why?)

**Comment by uablrek** [comment]
There are still some "swagger.json" files that referes the feature-gate `ServiceLBNodePortControl`;
```
kubernetes > grep -lr ServiceLBNodePortControl . | grep -Fv '.git' | grep -Fv _output
./staging/src/k8s.io/cli-runtime/artifacts/openapi/swagger.json
./staging/src/k8s.io/kubectl/testdata/openapi/swagger.json
./CHANGELOG/CHANGELOG-1.22.md
./pkg/features/kube_features.go
./vendor/sigs.k8s.io/kustomize/kyaml/openapi/kubernetesapi/v1212/swagger.json
```
I don't know how to get rid of them. They seem generated to I don't think I shall edit them.

When this is sorted out I will squash the commits with manual update but have "update generated" as a separate commit.

File: pkg/registry/core/service/storage/alloc.go (hunk 2 above)
Comment (thockin) [reply pkg/registry/core/service/storage/alloc.go:941]: pkg/apis/core/v1/defaults.go

That should be done before we get into `begin*()`, when we convert from versioned to internal.  Is that not what you are seeing?

Comment (aojea) [reply pkg/registry/core/service/storage/alloc.go:941]: yeah, I see that it is already defaulted, but I was trying to understand where, no probem then, this is fine, is mostly curiosity at this point
Thanks

Comment (aojea) [reply pkg/registry/core/service/storage/alloc.go:941]: for ref
image

**Comment by aojea** [comment]
/lgtm
maybe we should squash some changes?
Tim for approval
/assign @thockin

**Comment by uablrek** [comment]
> maybe we should squash some changes?

That was my intention, please see https://github.com/kubernetes/kubernetes/pull/107027#issuecomment-994961678.

But what about the remaining "swagger.json" files in `staging` and `vendor`? Is is ok anyway?

I posted a query on slack #kubernetes-contributors but have no answer so far.

**Comment by aojea** [comment]
> But what about the remaining "swagger.json" files in `staging` and `vendor`? Is is ok anyway?

_on generators we trust_ :smile: , if CI is happy :shrug:

**Comment by liggitt** [comment]
> > But what about the remaining "swagger.json" files in `staging` and `vendor`? Is is ok anyway?
> 
> _on generators we trust_ 😄 , if CI is happy 🤷
> 
> let's ask @liggitt

I'm not sure what the question is... I think all the swagger.json files under staging are test fixtures

**Comment by uablrek** [comment]
I made the 
```
grep -lr ServiceLBNodePortControl . | grep -Fv '.git' | grep -Fv _output
```
to make sure I flushed all references to the `ServiceLBNodePortControl` feature-gateway except in `./pkg/features/kube_features.go`. The "swagger.json" files are under version control. I think I have missed some generation script.

**Comment by uablrek** [comment]
Squashed manual updates. A separate commit for generated files.
The old comments are still in the "swagger.json" files as discussed above.

**Comment by uablrek** [comment]
Shall it be something in the "Does this PR introduce a user-facing change?" box in this PR?

**Comment by aojea** [comment]
> Shall it be something in the "Does this PR introduce a user-facing change?" box in this PR?

yeah, add a release note informing the feature move to GA

**Comment by sftim** [comment]
changelog entry nit:
```diff
-The `ServiceLBNodePortControl` graduates to GA. The feature-gate will be removed in 1.26.
+The `ServiceLBNodePortControl` feature graduates to GA. The feature gate will be removed in 1.26.
```

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Pull request merged**